### PR TITLE
Feature/storage metrics

### PR DESCRIPTION
**Pull request by ghost**
* Motivation
[Storage metrics and tests](https://github.com/comp413-2017/RDFS/issues/46)

* Description

added unfinished `storage/StorageMetrics`, and `test/storage/StorageTest`.
Implemented two core metrics - % space used and stdev of blocks per datanode.

The edits to the cmake files are good examples for adding a new package and its dependencies.

Currently, StorageTest doesn't fail because I just added a sample test. In the future we can add storage metric test cases to that file pretty easily.

### CODE REVIEW

**Comment by ghost** [comment]
also note that storage test setup is basically copied from deletetest. This means that instead of calling `rice-namenode/main.cc` to start up the namenode, we instead do it "manually" so we have a pointer to the zkWrapper.

File: test/integration/DeleteTest.cc [hunk 1]
```diff
@@ -26,13 +26,12 @@ class DeleteTest : public ::testing::Test {
     unsigned short port = 5351;
     nncli = new zkclient::ZkNnClient(zk);
     nncli->register_watches();
-    nn_translator = new ClientNamenodeTranslator(5351, nncli);
+    nn_translator = new ClientNamenodeTranslator(port, nncli);
```
Comment (adamnsm1) [inline test/integration/DeleteTest.cc:29]: Yay for code clarity

**Review by adamnsm1 — APPROVED** [review]
Everything looks good, dummy test works fine

**Review by LINKIWI — APPROVED** [review]
approval for build step changes

File: test/storage/StorageTest.cc [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      .detach();
+
+  ASSERT_EQ(0, system("python "
+                          "/home/vagrant/rdfs/test/integration/generate_file.py"
+                          " > expected_""testfile1234"));
+  // Put a file into rdfs.
+  system(
+      "hdfs dfs -fs hdfs://localhost:5351 -D dfs.blocksize=1048576 "
+          "-copyFromLocal expected_testfile1234 /f");
+  sleep(5);
```
Comment (jake-nyquist) [inline test/storage/StorageTest.cc:61]: Why does this magic number work? Just curious.

File: test/storage/StorageTest.cc [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+ */
+TEST_F(StorageTest, testExample) {
+  asio::io_service io_service;
+
+  RPCServer namenodeServer = nn_translator->getRPCServer();
+  std::thread(&RPCServer::serve, namenodeServer, std::ref(io_service))
+      .detach();
+
+  ASSERT_EQ(0, system("python "
+                          "/home/vagrant/rdfs/test/integration/generate_file.py"
```
Comment (jake-nyquist) [inline test/storage/StorageTest.cc:55]: Is there a way we could do a relative filepath here, it seems like this is brittle.

File: storage/source/StorageMetrics.cc [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      LOG(ERROR) << "Failed to get " << block_path;
+    }
+    dataNodeBlockCounts[i++] = static_cast<int>(block_ids.size());
+  }
+
+  return stDev(dataNodeBlockCounts);
+}
+
+float StorageMetrics::recoverySpeed() {
+  // TODO(ejd6): implement
```
Comment (jake-nyquist) [inline storage/source/StorageMetrics.cc:54]: lol

File: test/storage/StorageTest.cc [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    system(cmdLine.c_str());
+    sleep(3);
+  }
+
+  // Initialize and run the tests
+  ::testing::InitGoogleTest(&argc, argv);
+  int res = RUN_ALL_TESTS();
+
+  // Kill the DataNodes and NameNodes, and stop zookeeper
+  system("pkill -f StorageTestServer*");
```
Comment (jake-nyquist) [inline test/storage/StorageTest.cc:107]: If you run this test with other namenodes running on the machine, what happens? Seems like it kills everything.

**Review by jake-nyquist — APPROVED** [review]
This looks good to me with some very minor nits. Thanks for this!

File: storage/source/StorageMetrics.cc (hunk 4 above)
Comment (ghost) [reply storage/source/StorageMetrics.cc:54]: here we go

File: test/storage/StorageTest.cc (hunk 3 above)
Comment (ghost) [reply test/storage/StorageTest.cc:55]: i think we should keep it absolute. otherwise this call depends on the directory you run the test from

File: test/storage/StorageTest.cc (hunk 2 above)
Comment (ghost) [reply test/storage/StorageTest.cc:61]: honestly not sure if it's even necessary, i just copied the command from another test.

File: test/storage/StorageTest.cc (hunk 5 above)
Comment (ghost) [reply test/storage/StorageTest.cc:107]: the line after this kills any process matching "namenode" so i think it would kill any namenode running on the machine. in the context of test setup/teardown i think that's ok
line 107 only kills datanodes started in this test though, since other tests use different names.
